dubbo-common/src/main/java/org/apache/dubbo/common/URLBuilder.java

```diff
@@ -0,0 +1,342 @@
+            return this;
+        }
+        parameters.put(key, value);
+        return this;
+    }
+
+    public URLBuilder addParameters(Map<String, String> parameters) {
+        if (CollectionUtils.isEmptyMap(parameters)) {
+            return this;
+        }
+
+        boolean hasAndEqual = true;
+        for (Map.Entry<String, String> entry : parameters.entrySet()) {
+            String oldValue = this.parameters.get(entry.getKey());
+            String newValue = entry.getValue();
+            if (!Objects.equals(oldValue, newValue)) {
+                hasAndEqual = false;
+                break;
+            }
+        }
+        // return immediately if there's no change
+        if (hasAndEqual) {
+            return this;
+        }
+
+        this.parameters.putAll(parameters);
+        return this;
+    }
+
+    public URLBuilder addParametersIfAbsent(Map<String, String> parameters) {
+        if (CollectionUtils.isEmptyMap(parameters)) {
+            return this;
+        }
+        this.parameters.putAll(parameters);
+        return this;
+    }
+
+    public URLBuilder addParameters(String... pairs) {
+        if (pairs == null || pairs.length == 0) {
+            return this;
+        }
+        if (pairs.length % 2 != 0) {
+            throw new IllegalArgumentException("Map pairs can not be odd number.");
+        }
+        Map<String, String> map = new HashMap<>();
+        int len = pairs.length / 2;
+        for (int i = 0; i < len; i++) {
+            map.put(pairs[2 * i], pairs[2 * i + 1]);
+        }
+        return addParameters(map);
+    }
+
+    public URLBuilder addParameterString(String query) {
+        if (StringUtils.isEmpty(query)) {
+            return this;
+        }
+        return addParameters(StringUtils.parseQueryString(query));
+    }
+
+    public URLBuilder removeParameter(String key) {
+        if (StringUtils.isEmpty(key)) {
+            return this;
+        }
+        return removeParameters(key);
+    }
+
+    public URLBuilder removeParameters(Collection<String> keys) {
+        if (CollectionUtils.isEmpty(keys)) {
+            return this;
+        }
+        return removeParameters(keys.toArray(new String[0]));
+    }
+
+    public URLBuilder removeParameters(String... keys) {
+        if (keys == null || keys.length == 0) {
+            return this;
+        }
+        for (String key : keys) {
+            parameters.remove(key);
+        }
+        return this;
+    }
+
+    public URLBuilder clearParameters() {
+        parameters.clear();
+        return this;
+    }
+
+    public boolean hasParameter(String key) {
+        String value = getParameter(key);
+        return value != null && value.length() > 0;
+    }
+
+    public String getParameter(String key) {
+        String value = parameters.get(key);
+        if (StringUtils.isEmpty(value)) {
+            value = parameters.get(Constants.DEFAULT_KEY_PREFIX + key);
+        }
+        return value;
+    }
+}

```

Comment:
Does these two reading methods `hasParameter` `hasParameter` necessary for a Builder?

Reply:
The two methods are used inside the builder, and I just made them public because I think it's necessary because there are cases where we want to write code like this:
```java
URLBuilder builder = ...;
if (builder.hasParameter("serverSideProp")) {
    builder.addParameter("someServerSideProp1", v2);
    builder.addParameter("someServerSideProp2", v3);
}
```
without these methods we have to (again) create `URL`s just for calling the methods, what do you think

Reply:
Got it.